File: algorithms/[633]平方数之和/solution.py
# topics = ["数学"]

from typing import Set


class Solution:
    def judgeSquareSum(self, c: int) -> bool:
        """
        time O(sqrt(c)), space O(sqrt(c))
        """
        # 这里使用哈希表没有太大意义
        s: Set[int] = set()

        for i in range(c + 1):
            if i * i > c:
                break
            s.add(i * i)

        for t in s:
            if c - t in s:
                return True

        return False

File: algorithms/[633]平方数之和/solution.ts
// topics = ["数学"]

/**
 * sqrt 函数
 * time O(sqrt(c)), space O(1)
 */
function judgeSquareSum(c: number): boolean {
  for (let a = 0; a * a <= c; a += 1) {
    // sqrt 函数判断一个数是否是平方数
    const b = Math.sqrt(c - a * a);
    if (b == parseInt(b + '')) {
      return true;
    }
  }
  return false;
}

/**
 * 双指针
 * time O(sqrt(c)), space O(1)
 */
function judgeSquareSum2(c: number): boolean {
  let left = 0;
  let right = Math.floor(Math.sqrt(c));

  while (left <= right) {
    const sum = left * left + right * right;
    if (sum === c) {
      return true;
    } else if (sum > c) {
      right -= 1;
    } else {
      left += 1;
    }
  }

  return false;
}
